### ISSUE TITLE
materialize_meta_adapters returns 0 on peft >= 0.19 (adapters are real, not meta) — silent no-training risk

### ISSUE THREAD

**Issue by lesterppo** [MakazhanAlpamys/Soup#433]
## Summary

On peft ≥ 0.19.x, `get_peft_model` on a streaming meta skeleton creates the LoRA adapters as **real (non-meta) bf16 tensors**, not meta tensors. `materialize_meta_adapters` therefore returns **0** on that path — but for the opposite reason its docstring describes.

`materialize_meta_adapters` (`layer_stream_runtime.py`, docstring at ~1357-1363) documents that returning 0 means *"the optimizer happily accepts them and the run trains nothing"* — i.e. it assumes adapters are meta and that a 0 return means they were left without storage. On peft 0.19.x that assumption inverts: the adapters are already real, so `if not param.is_meta` skips them and the function returns 0 while the adapters *do* have storage and *do* train.

This is version-sensitive in the exact shape of #385/#387: the re-init path fires on peft 0.18.1 (meta adapters) and silently does nothing on 0.19.x (real adapters), so nothing in CI (pinned to 0.18.1) exercises the newer behaviour.

## Evidence

Environment (free Colab T4, 2026-08-16):

- `torch 2.11.0+cu128` (CUDA 12.8), `peft 0.19.1` (from the saved `adapter_config.json`), Python 3.12.13.

At the crash point (first `clip_grad_norm_`), a dump of trainable params:

```
{'torch.bfloat16': 128}
base_model.model.model.layers.0.inner.self_attn.q_proj.lora_A.default.weight  (8, 4096)
...
```

Two observations:

1. **dtype is bf16** — the `autocast_adapter_dtype` upcast to fp32 that peft 0.18.1 performs in `PeftModel.__init__` → `_cast_adapter_dtype` does not happen on this path on 0.19.1. This is the direct cause of the fp16 GradScaler crash (#425).
2. **non-meta** — inferred, not directly dumped: with #429's cast applied the same run completes and writes an adapter with **128/128 tensors non-zero**. A meta param has no storage and cannot be updated by the optimizer, so a 128/128-non-zero adapter is proof the adapters were real (non-meta) at build time.

## Why this matters beyond #425

The crash (#425) announces itself. This one would not: the "adapters are meta but `materialize_meta_adapters` skipped them" failure — the case the docstring warns about — is silent no-training, and on 0.19.x the function now returns 0 as a matter of course, so its return value can no longer be used to detect that failure. If a future peft version (or a different streaming path) produces meta adapters that the `is_meta` guard misses, the run would train nothing with a healthy loss curve.

## Suggested direction (not implemented here)

Make the streaming build assert on the post-`get_peft_model` adapter state instead of trusting `materialize_meta_adapters`' return: after the build, verify every trainable `*lora_*` param has real storage (not `is_meta`) and fail loudly otherwise. Optionally make `materialize_meta_adapters` distinguish "no meta adapters to re-init (already real)" from "meta adapters that were skipped".

The dtype half of this is fixed by #429 (cast bf16 → fp32); this issue tracks the *meta/non-meta* half, which is independent.

**Comment by MakazhanAlpamys** [comment]
Good issue, and thank you for filing it separately rather than letting it ride inside #429 — that is the right instinct, and the reasoning in it is sound.

Two things I want to record, because they change how the issue should be read.

## Your inference is valid, and I want to say why explicitly

You flagged the non-meta half as **inferred, not dumped**, which is the correct label. But the inference is not weak: a meta parameter has no storage and cannot be written by an optimizer, so **128/128 non-zero tensors in the saved adapter is proof of real storage at build time**. That is a stronger argument than a direct `is_meta` print would have been, because it observes the property through its consequence rather than through a flag that could itself be misread.

Flagging it as inferred anyway is exactly the standard this project asks for. Noting it so nobody later downgrades the finding for being labelled honestly.

## Sharpening the framing

The title says "silent no-training risk", and I would narrow that by one step: **on your stack, training worked**. Nothing is broken today. What broke is that a **guard silently stopped being a guard**.

`materialize_meta_adapters` returning 0 used to mean "there were no meta adapters to re-init". On peft ≥ 0.19 it means that *as a matter of course*, so the return value can no longer distinguish:

- "adapters are already real — nothing to do" (benign, your case), from
- "adapters are meta and the `is_meta` check missed them" (the silent no-training case the docstring warns about).

Those two are now indistinguishable at the call site. That is a real defect and worth fixing, but it is a **loss of detection**, not an active failure — and stating it that way makes it a clearer target.

It also puts it in a family this project keeps meeting: a signal that is correct today because of a coincidence upstream. Compare #385/#387, where `torch.cuda.is_bf16_supported()` answers `True` on a T4 through emulation, so the first fix was inert on the exact hardware it was written for. Same shape: the check is not wrong, it just stopped discriminating.

## On your suggested direction — agreed, with one addition

Asserting on the post-`get_peft_model` state rather than trusting a return value is right. One thing to add: whatever assertion lands needs a **control that fails**. A check that every trainable `*lora_*` param has real storage will pass on both peft versions today, so on its own it cannot demonstrate it would catch the bad case. It needs a companion that forces a meta adapter through and confirms the assertion fires — otherwise we replace an inert guard with an untested one.

Making `materialize_meta_adapters` return a three-way answer rather than a count is also worth considering, and there is precedent in this codebase: `StepPeak` in `layer_stream_runtime.py` deliberately has **three** outcomes rather than two, precisely so "it did not happen" cannot be confused with "it failed".

## Practicalities

Labelled `bug` + `help wanted`. Not blocking #429 — the dtype half stands on its own, and as you say the cast is a no-op on 0.18.1 where adapters are already fp32.

If you get another Colab session, the one measurement that would close the remaining gap is a direct `[(n, p.is_meta) for n, p in model.named_parameters() if p.requires_grad][:5]` right after `get_peft_model` on 0.19.x. Not needed to accept the issue — just the difference between inferred and measured, and this repo tries to keep that line visible.

**Comment by Faisal01011** [comment]
I’d like to work on this. My plan is to add a post-materialization invariant that verifies every trainable `lora_*` parameter has real storage, plus a negative regression test that deliberately leaves a meta adapter and proves the guard fails. Could you assign this issue to me?

**Comment by MakazhanAlpamys** [comment]
Yours — go ahead, and your plan is right. The negative regression test is the part that matters most, and you named it yourself without being asked.

Three things to know before you start, because this one is a real step up from the recipe work and the traps are not obvious.

## 1. You probably cannot reproduce the trigger locally

This only manifests on **peft ≥ 0.19**, and this repo pins **0.18.1** — that is precisely why it survived. On 0.18.1 the adapters arrive meta and fp32; on 0.19.1 they arrive real and bf16. So if you write a test that needs the newer peft to fail, **CI will never run it**, and you will have replaced an inert guard with an untested one.

The way through is the same one `tests/test_issue385_stream_dtype.py` uses, and its docstring says so explicitly: *assert the DECISION, with the capability stubbed.* Build the meta adapter synthetically with fakes rather than by installing a peft that produces one. Then the test works on both versions and on any machine.

## 2. Your invariant will pass on both peft versions today — so prove it can fail

"Every trainable `lora_*` param has real storage" is **true right now** on 0.18.1 (materialized) and on 0.19.1 (never meta). A check that cannot currently fail looks exactly like a check that works.

That is what your negative test is for, and it is not optional here: force a meta `lora_*` param through and assert the guard fires, naming it. Without that control, the invariant is decoration — and decoration is what this issue is about, since `materialize_meta_adapters` returning 0 is itself a signal that stopped discriminating.

A worked example of the same discipline landed today: in #429 the scanner test was proven real two ways — by removing a call from an existing trainer *and* by dropping in a brand-new fake trainer file. Both made it fail, naming the file. That is the bar.

## 3. Consider a three-way answer, not a boolean

`materialize_meta_adapters` currently returns a count, and the whole defect is that `0` now conflates "nothing to do" with "skipped something it should have caught". If you touch that function, `StepPeak` in `utils/layer_stream_runtime.py` is the house precedent — it deliberately has **three** outcomes rather than two, so "it did not happen" cannot be mistaken for "it failed".

## Housekeeping

GitHub will not let me set you as assignee (collaborators only — nothing to fix on your side), so this comment is the claim of record: **#433 is yours.**

[@lesterppo](https://github.com/lesterppo) filed it and wrote "suggested direction (not implemented here)", so I am reading that as not planning to implement — but flagging it so they can say otherwise. They also have the peft 0.19.1 environment, which neither of us does, so if you land the invariant they are the natural person to confirm it behaves on the stack where the bug actually appears.

One honest note: the remaining measurement gap in the issue is a direct `[(n, p.is_meta) for n, p in model.named_parameters() if p.requires_grad][:5]` right after `get_peft_model` on 0.19.x. Nobody has captured that yet — the non-meta conclusion is inferred (soundly) from 128/128 non-zero adapter tensors. If your work does not depend on it, say so; if it does, ask for it rather than assuming.

**Comment by MakazhanAlpamys** [comment]
Closed by #435 — thanks to **[@Faisal01011](https://github.com/Faisal01011)** (Faisal Fayaz), and to **[@lesterppo](https://github.com/lesterppo)** who found and filed it.

Merged as [`66a3c0a`](https://github.com/MakazhanAlpamys/Soup/commit/66a3c0a).

## What shipped

The count returned by `materialize_meta_adapters` is now explicitly a **diagnostic** — its docstring says "not a success signal… zero is healthy" — and the decision moved into a separate postcondition, `assert_trainable_adapters_materialized`, called from `build_streamed_model`.

That is a deliberate divergence from the three-valued return I suggested, and the better call: with a separate postcondition the caller can no longer *choose* to ignore it, which a richer return value would still permit.

**The strongest property, which the PR under-claims:** the guard's predicate is the *same* predicate `materialize_meta_adapters` selects on, plus `requires_grad`. So it is a true postcondition — it structurally cannot refuse a build the materializer could have repaired.

## Verified by mutation, since that was the whole risk here

This issue is about a guard that silently stopped discriminating, so a fix that is itself unguarded would have been the same defect twice. It is not:

- Deleting the call from `build_streamed_model` → suite red, `DID NOT RAISE RuntimeError`.
- Six mutations, five killed — including stripping the parameter name from the message (both `match=r"lora_A"` assertions fire) and moving the guard ahead of `get_peft_model`, which verifies a claim the test's own docstring makes rather than leaving it aspirational.

The trigger needs peft ≥ 0.19 and this repo pins 0.18.1, so the tests stub the capability and assert the **decision** — the `test_issue385_stream_dtype.py` discipline. Five tests, zero skips, CPU-only.

Behaviour confirmed in both directions independently of the PR's own tests: fires on a genuinely meta adapter naming both parameters, quiet when peft hands over real tensors (`materialize` returns 0) and quiet after a 0.18.1-style materialisation (returns 2). So no peft 0.19 user gets a spurious failure.

## Known gap, non-blocking

Dropping `"lora_" in name` from the filter survives the suite, so the restriction the docstring calls deliberate is unpinned. Inert on today's path — `build_streamed_model` freezes every base parameter before `get_peft_model` — and the risk direction is *spurious refusal* rather than silent no-training, i.e. the opposite of this issue.

## Still open, and worth stating

The original measurement gap remains: nobody has captured `[(n, p.is_meta) for n, p in model.named_parameters() if p.requires_grad][:5]` right after `get_peft_model` on peft 0.19.x. The non-meta conclusion is *inferred* — soundly, from 128/128 non-zero adapter tensors — but inferred.

The guard is correct either way and does not depend on it. Recording it so the line between inferred and measured stays visible.
